crates/tests-e2e/src/test_subscriber.rs

```diff
@@ -0,0 +1,140 @@
+use std::sync::Arc;
+use std::sync::Mutex;
+
+use coordinator_commons::TradeParams;
+use native::api::WalletInfo;
+use native::event::subscriber::Subscriber;
+use native::event::EventType;
+use native::trade::order::Order;
+use native::trade::position::Position;
+use tokio::sync::watch;
+
+pub struct Senders {
+    wallet_info: watch::Sender<Option<WalletInfo>>,
+    order: watch::Sender<Option<Order>>,
+    order_filled: watch::Sender<Option<Box<TradeParams>>>,
+    position: watch::Sender<Option<Position>>,
+}
+
+// Subscribes to events destined for the frontend (typically Flutter app) and
+// provides a convenient way to access the current state.
+pub struct TestSubscriber {
+    wallet_info: watch::Receiver<Option<WalletInfo>>,
+    order: watch::Receiver<Option<Order>>,
+    order_filled: watch::Receiver<Option<Box<TradeParams>>>,
+    position: watch::Receiver<Option<Position>>,
+    // TODO add prices and close position watches
+}
+
+impl TestSubscriber {
+    pub fn new() -> (Self, ThreadSafeSenders) {
+        let (wallet_info_tx, wallet_info_rx) = watch::channel(None);
+        let (order_tx, order_rx) = watch::channel(None);
+        let (order_filled_tx, order_filled_rx) = watch::channel(None);
+        let (position_tx, position_rx) = watch::channel(None);
+
+        let senders = Senders {
+            wallet_info: wallet_info_tx,
+            order: order_tx,
+            order_filled: order_filled_tx,
+            position: position_tx,
+        };
+
+        let rx = Self {
+            wallet_info: wallet_info_rx,
+            order_filled: order_filled_rx,
+            order: order_rx,
+            position: position_rx,
+        };
+        (rx, ThreadSafeSenders(Arc::new(Mutex::new(senders))))
+    }
+
+    pub fn wallet_info(&self) -> Option<WalletInfo> {
+        self.wallet_info.borrow().as_ref().cloned()
+    }
+
+    pub fn order(&self) -> Option<Order> {
+        self.order.borrow().as_ref().copied()
+    }
+
+    pub fn order_filled(&self) -> Option<Box<TradeParams>> {
+        self.order_filled.borrow().as_ref().cloned()
+    }
+
+    pub fn position(&self) -> Option<Position> {
+        self.position.borrow().as_ref().cloned()
+    }
+}
+
+impl Subscriber for Senders {
+    fn notify(&self, event: &native::event::EventInternal) {
+        dbg!("ZOSIA");
```

Comment:
Hahaha.

Reply:
yeah, that's just my internal debugging  😅

 shoutout to my friend from Nokia who always used this token to find it in logs all the years ago